client/src/objects/drawin.rs

```diff
@@ -4,31 +4,93 @@
 // NOTE need to store the drawable in lua, because it's a reference to a
 // drawable a lua object
 
-use std::convert::TryFrom;
-
-use cairo::ImageSurface;
 use rlua::{self, prelude::LuaInteger, Context, Table, ToLua, UserData, UserDataMethods, Value};
 
-use crate::area::{Area, Margin, Origin, Size};
-use crate::common::{
-    class::{self, Class, ClassBuilder},
-    object::{self, Object, ObjectBuilder},
-    property::Property
+use crate::{
+    area::{Area, Margin, Origin, Size},
+    common::{
+        class::{self, Class, ClassBuilder},
+        object::{self, Object, ObjectBuilder},
+        property::Property
+    },
+    wayland_obj::{create_layer_surface, Buffer, LayerSurface}
 };
-use crate::objects::drawable::Drawable;
+
+use super::drawable::Drawable;
 
 pub const DRAWINS_HANDLE: &'static str = "__drawins";
 
-#[derive(Debug, Default)]
+#[derive(Default)]
 pub struct DrawinState {
-    // Note that the drawable is stored in Lua.
-    // TODO WINDOW_OBJECT_HEADER??
-    ontop: bool,
+    // ontop: bool,
     visible: bool,
-    cursor: String,
+    // cursor: String,
+    // drawable : Drawable // NOTE: this is stored in lua!
     geometry: Area,
+    /// Do we have a pending geometry change that still needs to be applied?
     geometry_dirty: bool,
-    surface: Option<ImageSurface>
+    // surface: Option<ImageSurface>, // TODO(ried): remove
+
+    // TODO WINDOW_OBJECT_HEADER??
+    surface: Option<LayerSurface>, // window : Window,
+    // frame_window : Window,
+    // opacity : double,
+    struts: Margin
+}
+// buttons : Vec<Button>,
+// border_needs_update : bool,
+// border_color : Color,
+// border_width : u16,
+// type : WindowType,
+// border_width_callback : Fn(_ , old_width : u16, new_width : u16)
+
+impl DrawinState {
+    /// Update the geometry of this drawin and return the signals to be emitted.
```

Comment:
Is there a reason we send the signals at the end all at once instead of just doing it one by one?

Reply:
I will remove the update-logic from the ..State objects, then the signal can be emitted one-by-one.